From 11e19044efff6808d98bf637f2e8d35681a75365 Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Mon, 2 Apr 2018 09:18:58 +0700
Subject: [PATCH 06/25] user.exe16: Implement CountThaiCharType.
To: wine-devel <wine-devel@winehq.org>

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>
---
 dlls/user.exe16/thai.c | 25 +++++++++++++++++++++++--
 1 file changed, 23 insertions(+), 2 deletions(-)

diff --git a/dlls/user.exe16/thai.c b/dlls/user.exe16/thai.c
index 575a23d008..f19fd95ad1 100644
--- a/dlls/user.exe16/thai.c
+++ b/dlls/user.exe16/thai.c
@@ -258,8 +258,29 @@ WORD WINAPI ValidateThaiStr16( SEGPTR lpSrc, SEGPTR lpDest, WORD nCount, BYTE cC
  */
 WORD WINAPI CountThaiCharType16( SEGPTR lpString, WORD nCount, WORD Ntype )
 {
-    FIXME("(%p, %u, %04x), stub\n", (void*)&lpString, nCount, Ntype);
-    return 0;
+    int i;
+    WORD count = 0;
+    char *ptr = MapSL(lpString);
+    int lenght = strlen(ptr);
+
+    TRACE("(\"%s\", %u, %04x)\n", ptr, nCount, Ntype);
+
+    for(i = 0; i < nCount; i++)
+    {
+        if( ( nCount > lenght ) && ( *ptr == '\0' ) )
+            break;
+
+        if( ThaiCharType16( *ptr ) & Ntype )
+        {
+            count++;
+        }
+
+        ptr++;
+    }
+
+    TRACE("count %d\n", count);
+
+    return count;
 }
 
 
-- 
2.22.0

