From 42a52227d857e5f17a95380a8bdbaa29f6137b93 Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Mon, 2 Apr 2018 10:07:31 +0700
Subject: [PATCH 09/25] user.exe16: Fix default profile string.
To: wine-devel <wine-devel@winehq.org>

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>
---
 dlls/user.exe16/thai.c | 6 +++---
 1 file changed, 3 insertions(+), 3 deletions(-)

diff --git a/dlls/user.exe16/thai.c b/dlls/user.exe16/thai.c
index eb32634dbe..99f6192f6c 100644
--- a/dlls/user.exe16/thai.c
+++ b/dlls/user.exe16/thai.c
@@ -346,7 +346,7 @@ SEGPTR WINAPI lstrDayName16( WORD nDay, SEGPTR lpDest, WORD nFormat )
             if(iThaiDate)
                 /* load setting from WIN.INI
                 * Default is "dddd'ที่ 'd' 'MMMM' พ.ศ. 'yyyy" */
-                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB8\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
+                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB7\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
                                             (LPSTR)szDateForm, sizeof(szDateForm) );
             else
                 iLenght = GetProfileStringA( "intl", "sLongDate", "dd MMMM yyyy",
@@ -466,7 +466,7 @@ SEGPTR WINAPI lstrMonthName16( WORD nMonth, SEGPTR lpDest, WORD nFormat )
             if(iThaiDate)
                 /* load setting from WIN.INI
                 * Default is "dddd'ที่ 'd' 'MMMM' พ.ศ. 'yyyy" */
-                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB8\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
+                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB7\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
                                             (LPSTR)szDateForm, sizeof(szDateForm) );
             else
                 iLenght = GetProfileStringA( "intl", "sLongDate", "dd MMMM yyyy",
@@ -544,7 +544,7 @@ SEGPTR WINAPI lstrYearNumber16( WORD nYear, SEGPTR lpDest, WORD nFormat )
             if(iThaiDate)
                 /* load setting from WIN.INI
                 * Default is "dddd'ที่ 'd' 'MMMM' พ.ศ. 'yyyy" */
-                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB8\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
+                iLenght = GetProfileStringA( "Thai", "sLongDate", "dddd'\xB7\xD5\xE8 'd' 'MMMM' \xBE.\xC8. 'yyyy",
                                             (LPSTR)szDateForm, sizeof(szDateForm) );
             else
                 iLenght = GetProfileStringA( "intl", "sLongDate", "dd MMMM yyyy",
-- 
2.22.0

