From 9d2ed5a2b8a6ab915608d11861fb36b3ca06017c Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Wed, 4 Apr 2018 11:51:46 +0700
Subject: [PATCH 13/26] user.exe16: Correct return value for SetTaskInterface.

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>
---
 dlls/user.exe16/thai.c | 11 ++++++-----
 1 file changed, 6 insertions(+), 5 deletions(-)

diff --git a/dlls/user.exe16/thai.c b/dlls/user.exe16/thai.c
index cfa3c385a88..8ae1bf112ef 100644
--- a/dlls/user.exe16/thai.c
+++ b/dlls/user.exe16/thai.c
@@ -31,7 +31,7 @@
 WINE_DEFAULT_DEBUG_CHANNEL(user);
 
 /* Thai API */
-static WORD THAI_Interface = 0;         /* Thai API interface flags */
+static WORD THAI_Interface = STANDARD_INTERFACE;    /* Thai API interface flags */
 static WORD THAI_KeyboardLang = KB_ENG; /* Thai API keyboard language */
 static BOOL16 THAI_KbdSeqCheck = TRUE;  /* Thai API keyboard sequence check flags */
 
@@ -621,9 +621,10 @@ BOOL16 WINAPI SetThaiKbdSeqCheck16( BOOL16 fCheck )
 /***********************************************************************
  *           SetTaskInterface (USER.498)
  */
-WORD WINAPI SetTaskInterface16( HTASK16 hTask, WORD dwFlags )
+WORD WINAPI SetTaskInterface16( HTASK16 hTask, WORD wMode )
 {
-    FIXME("(%p, %04x), semi-stub\n", (void*)&hTask, dwFlags);
-    THAI_Interface = dwFlags;
-    return 1;
+    WORD oldInterface = THAI_Interface;
+    FIXME("(%p, %04x), semi-stub\n", (void*)&hTask, wMode);
+    THAI_Interface = wMode;
+    return oldInterface;
 }
-- 
2.39.0

