From 5b4975660c39cc26855cd4fc20710fc65ff75d9b Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Wed, 11 Apr 2018 23:02:49 +0700
Subject: [PATCH 22/26] user.exe16: Fix SetThaiKbdSeqCheck return value.

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>
---
 dlls/user.exe16/thai.c | 3 ++-
 1 file changed, 2 insertions(+), 1 deletion(-)

diff --git a/dlls/user.exe16/thai.c b/dlls/user.exe16/thai.c
index 12def7e4b0..ea63b750d5 100644
--- a/dlls/user.exe16/thai.c
+++ b/dlls/user.exe16/thai.c
@@ -715,8 +715,9 @@ BOOL16 WINAPI IsThaiKbdSeqCheck16(void)
  */
 BOOL16 WINAPI SetThaiKbdSeqCheck16( BOOL16 fCheck )
 {
+    BOOL16 oldFlag = THAI_KbdSeqCheck;
     THAI_KbdSeqCheck = fCheck;
-    return TRUE;
+    return oldFlag;
 }
 
 
-- 
2.27.0

