From 9a85fe6efe270b896c481679611a76d83c0bb6fc Mon Sep 17 00:00:00 2001
From: Khralkatorrix <garuda2550@gmail.com>
Date: Wed, 10 Jun 2020 00:54:48 +0700
Subject: [PATCH 25/26] readme: Add readme.

Signed-off-by: Khralkatorrix <garuda2550@gmail.com>
---
 README.thaiapi | 107 +++++++++++++++++++++++++++++++++++++++++++++++++
 1 file changed, 107 insertions(+)
 create mode 100644 README.thaiapi

diff --git a/README.thaiapi b/README.thaiapi
new file mode 100644
index 00000000000..0172f35ae88
--- /dev/null
+++ b/README.thaiapi
@@ -0,0 +1,107 @@
+# Wine Thai API
+
+Wine with support for Thai API from Windows 3.0/3.1/95/98/ME Thai Edition.
+
+# Thai API status
+
+    * = to be implement
+    x = don't implement
+
+| Status      | API         | Function                 | Note                         |
+| ----------- |:----------- | :----------------------- | :--------------------------- |
+| stub        | KEYBOARD.8  | TkEnable                 | Win 3.0a Thai + Win 3.1 Thai |
+| stub        | KEYBOARD.9  | TkDisable                | Win 3.0a Thai + Win 3.1 Thai |
+| implemented | KEYBOARD.10 | IsThaiExist              | Win 3.0a Thai + Win 3.1 Thai |
+| implemented | KEYBOARD.11 | GetThaiStatus            | Win 3.0a Thai is GetThaiStatus, Win 3.1 Thai is WEP |
+| implemented | KEYBOARD.12 | SetThaiStatus            | Win 3.0a Thai is SetThaiStatus, Win 3.1 Thai is NotifyLanguage |
+| stub        | KEYBOARD.13 | SetCheckSequence         | Win 3.0a Thai is SetCheckSequence, Win 3.1 Thai is ___EXPORTEDSTUB |
+| implemented | KEYBOARD.15 | SetThaiKbdLayout         | |
+| implemented | KEYBOARD.16 | GetThaiKbdLayout         | |
+| implemented | KEYBOARD.17 | lstrKbdThaiToEng         | |
+| implemented | KEYBOARD.18 | lstrKbdEngToThai         | |
+| implemented | KEYBOARD.19 | lstrNumThaiToArabic      | |
+| implemented | KEYBOARD.20 | lstrNumArabicToThai      | |
+|x stub       | KEYBOARD.21 | SetBahtKey               | |
+|x stub       | KEYBOARD.23 | SetThaiCapsLockEffect    | |
+| implemented | GDI.490     | IsThaiFont               | Conflict with CloseEnhMetafile, required by Microsoft Word 6.0 Thai Edition |
+|x stub       | GDI.491     | IsInvalidDisplaySeqBeep  | Conflict with CopyEnhMetafile |
+|x stub       | GDI.492     | SetInvalidDisplaySeqBeep | Conflict with CreateEnhMetafile |
+| semi-stub   | USER.474    | GetTaskInterface         | |
+| stub        | USER.475    | GetKeyboardLanguage      | Conflict with SetScrollInfo |
+| stub        | USER.476    | SetKeyboardLanguage      | Conflict with GetScrollInfo, **FIXME: Properly implement this function** |
+| stub        | USER.477    | SetLanguageToggleKey     | Conflict with GetKeyboardLayoutName |
+|* stub       | USER.485    | FindThaiWordBreak        | use cttex, libthai, etc.? |
+| implemented | USER.486    | ThaiCharType             | |
+|x stub       | USER.487    | IsInvalidKbdSeqBeep      | |
+|x stub       | USER.488    | SetInvalidKbdSeqBeep     | |
+|* stub       | USER.489    | IsThaiCharCombinable     | |
+|* stub       | USER.490    | ThaiStrValid             | use IsThaiCharCombinable to check? |
+|* stub       | USER.491    | ValidateThaiStr          | |
+| implemented | USER.492    | CountThaiCharType        | |
+| implemented | USER.493    | lstrDayName              | |
+| implemented | USER.494    | lstrMonthName            | |
+| implemented | USER.495    | lstrYearNumber           | |
+| implemented | USER.496    | IsThaiKbdSeqCheck        | |
+| implemented | USER.497    | SetThaiKbdSeqCheck       | |
+| semi-stub   | USER.498    | SetTaskInterface         | |
+| stub        | LPKTHAI.DLL | | Dummy dll, required by Microsoft Office 97 Thai Edition Language Switcher. |
+
+### Note:
+- Other than TkEnable, TkDisable, IsThaiExist, GetThaiStatus, SetThaiStatus
+and SetCheckSequence function in keyboard.drv, all other Thai API functions
+in GDI and USER is not exist in Windows 3.0a Thai Edition.
+- In oder for Thai functions to function correctly, you have to add this
+entry in win.ini in your wine prefix, must use TIS620 character encoding.
+
+    [Thai]
+    iThaiDate=1
+    sLongDate=dddd'ที่ 'd' 'MMMM' พ.ศ. 'yyyy
+
+### Will not implement:
+- Baht to Mai-eak Mai-to key
+
+### Todo:
+- Implement IsThaiCharCombinable
+- Implement ThaiStrValid
+- Implement ValidateThaiStr
+- Implement FindThaiWordBreak, use cttex or libthai or ...
+- Process message `WM_LANGUAGE_CHANGE` in user.exe16/message.c:DefWindowProc16 to change keyboard language. Keyboard language is changed to the one that was given in wParam.
+- Win 3.1 Thai use 0xEF to sperate English and Thai text on menu and captions,
+modify the display text function to support this behavior?
+example: [english str][0xEF][thai str]
+- SetKeyboardLanguage, use change keyboard language function in Win32
+- Modify ToAscii to support Thai input
+- Fix display string crash with Thai text
+- Add Thai bitmap system font
+- In ThaiCharType16 function, consider using lookup table
+- Invisigate Thai API usage in Windows 95 Thai Edition and Microsoft Office 95 Thai Edition
+
+### List of software that use Thai API:
+- CU-Writer for Windows (CWW77)
+CWW.EXE
+API: SetThaiKbdLayout, GetThaiKbdLayout, lstrKbdThaiToEng, SetThaiKbdSeqCheck
+CWWPRINT.EXE
+API: SetThaiKbdSeqCheck, GetKeyboardLanguage, IsThaiKbdSeqCheck, SetKeyboardLanguage, SetLanguageToggleKey, GetThaiKbdLayout, lstrKbdThaiToEng
+- iRC ZeWrite Professional for Windows 1.0
+API: FindThaiWordBreak, lstrKbdEngToThai, lstrKbdThaiToEng, SetThaiKbdSeqCheck, ThaiCharType
+- Lotus Approach 3.0 Thai?
+- Lotus ccMail 2.1 Thai?
+- Lotus SmartSuite Release 2.0 Thai?
+- Lotus SmartSuite Release 3.0 Thai?
+- Microsoft Excel 5.0a Thai Edition
+API: FindThaiWordBreak, GetKeyboardLanguage, IsThaiCharCombinable, SetKeyboardLanguage
+- Microsoft Office 4.2 Standard Thai Edition and Microsoft Office 4.3 Professional Thai Edition
+Only Excel 5.0a, Word 6.0a is using Thai API.
+- Microsoft Office 95 Thai Edition?
+- Microsoft Office 97 Professional Thai Edition
+Language Switcher require `LPKTHAI.DLL` to switch to Thai language.
+- Microsoft Mail for Windows 3.02 (Bundled with Windows for Workgroup 3.1 Thai Edition)
+API: FindThaiWordBreak, lstrKbdThaiToEng, lstrKbdEngToThai , lstrDayName, lstrMonthName, lstrYearNumber, GetTaskInterface, SetTaskInterface
+- Microsoft Schedule+ for Windows 1.0 (Bundled with Windows for Workgroup 3.1 Thai Edition)
+API: FindThaiWordBreak, lstrKbdThaiToEng, lstrKbdEngToThai, GetTaskInterface
+- Microsoft Windows 3.1 Thai Edition NOTEPAD.EXE
+API: GetTaskInterface, SetTaskInterface
+- Microsoft Windows 3.1 Thai Edition TWRITE.EXE
+API: FindThaiWordBreak
+- Microsoft Word 6.0a Thai Edition
+API: IsThaiFont
-- 
2.39.0

