From 5042c3a969c6ad9903ba5e2489f616a2af98af53 Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Mon, 18 Feb 2019 23:27:32 +0700
Subject: shell.dll16: Port the patch from winevdm.

https://github.com/otya128/winevdm/tree/master/shell

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>

diff --git a/dlls/shell.dll16/shell.c b/dlls/shell.dll16/shell.c
index be8ffeb817..38fc6fd77f 100644
--- a/dlls/shell.dll16/shell.c
+++ b/dlls/shell.dll16/shell.c
@@ -74,6 +74,7 @@ static UINT	uMsgShellActivate = 0;
 
 static HICON convert_icon_to_32( HICON16 icon16 )
 {
+    if (!icon16) return 0;
     CURSORICONINFO *info = GlobalLock16( icon16 );
     void *and_bits = info + 1;
     void *xor_bits = (BYTE *)and_bits + info->nHeight * 2 * ((info->nWidth + 15) / 16);
@@ -85,6 +86,7 @@ static HICON convert_icon_to_32( HICON16 icon16 )
 
 static HICON16 convert_icon_to_16( HINSTANCE16 inst, HICON icon )
 {
+    if (!icon) return 0;
     static HICON16 (WINAPI *pCreateIcon16)(HINSTANCE16,INT16,INT16,BYTE,BYTE,LPCVOID,LPCVOID);
     ICONINFO info;
     BITMAP bm;
