From 3c962cba0deaf9afa1c0b0f955c465152d1370a6 Mon Sep 17 00:00:00 2001
From: Kai Krakow <kai@kaishome.de>
Date: Wed, 26 Sep 2018 21:06:07 +0200
Subject: ntdll: Enable STAGING_SHARED_MEMORY by default

Signed-off-by: Kai Krakow <kai@kaishome.de>

diff --git a/dlls/ntdll/server.c b/dlls/ntdll/server.c
index 3d88b47a744..56f3c1e9026 100644
--- a/dlls/ntdll/server.c
+++ b/dlls/ntdll/server.c
@@ -1222,7 +1222,7 @@ static inline BOOL experimental_SHARED_MEMORY( void )
     if (enabled == -1)
     {
         const char *str = getenv( "STAGING_SHARED_MEMORY" );
-        enabled = str && (atoi(str) != 0);
+        enabled = !str || (atoi(str) != 0);
     }
     return enabled;
 }
