From d8bd8dc9038e3600f8ef888de60a8a2934f3351f Mon Sep 17 00:00:00 2001
From: Khral Steelforge <garuda2550@gmail.com>
Date: Sun, 19 Aug 2018 15:04:34 +0700
Subject: ole32: Fix WIC error in Path of Exile.

From https://bugs.winehq.org/show_bug.cgi?id=42695

Signed-off-by: Khral Steelforge <garuda2550@gmail.com>

diff --git a/dlls/ole32/compobj.c b/dlls/ole32/compobj.c
index bd28f42b33b..37d4aa7e7e0 100644
--- a/dlls/ole32/compobj.c
+++ b/dlls/ole32/compobj.c
@@ -3444,17 +3444,23 @@ HRESULT WINAPI DECLSPEC_HOTPATCH CoCreateInstanceEx(
 
     init_multi_qi(cmq, pResults, E_NOINTERFACE);
 
-    hres = CoGetTreatAsClass(rclsid, &clsid);
-    if(FAILED(hres))
-        clsid = *rclsid;
-
     if (!(apt = apartment_get_current_or_mta()))
     {
-        ERR("apartment not initialised\n");
-        return CO_E_NOTINITIALIZED;
+        WARN("apartment not initialised; initializing now\n");
+        CoInitialize(NULL);
+
+        if (!(apt = apartment_get_current_or_mta()))
+        {
+            ERR("apartment still not initialised\n");
+            return CO_E_NOTINITIALIZED;
+        }
     }
     apartment_release(apt);
 
+    hres = CoGetTreatAsClass(rclsid, &clsid);
+    if(FAILED(hres))
+        clsid = *rclsid;
+
     /*
      * The Standard Global Interface Table (GIT) object is a process-wide singleton.
      */
