From 455786a3e6750ae502376da06e41424d012d52bf Mon Sep 17 00:00:00 2001
From: Zebediah Figura <zfigura@codeweavers.com>
Date: Mon, 16 Sep 2019 12:11:36 -0500
Subject: [PATCH] ntdll: fix for spincounts

---
 dlls/ntdll/fsync.c | 14 +++++---------
 1 file changed, 5 insertions(+), 9 deletions(-)

diff --git a/dlls/ntdll/fsync.c b/dlls/ntdll/fsync.c
index b3871ad3834..2c4c656d276 100644
--- a/dlls/ntdll/fsync.c
+++ b/dlls/ntdll/fsync.c
@@ -860,18 +860,14 @@ static NTSTATUS __fsync_wait_objects( DWORD count, const HANDLE *handles,
 
                         for (spin = 0; spin < spincount + 1; ++spin)
                         {
-                            do
+                            if ((current = __atomic_load_n( &semaphore->count, __ATOMIC_SEQ_CST ))
+                                    && __sync_val_compare_and_swap( &semaphore->count, current, current - 1) == current)
                             {
-                                if (!(current = semaphore->count)) goto out;
-                            } while (__sync_val_compare_and_swap( &semaphore->count, current, current - 1 ) != current);
+                                TRACE("Woken up by handle %p [%d].\n", handles[i], i);
+                                return i;
+                            }
                             small_pause();
                         }
-out:
-                        if (current)
-                        {
-                            TRACE("Woken up by handle %p [%d].\n", handles[i], i);
-                            return i;
-                        }
 
                         futexes[i].addr = &semaphore->count;
                         futexes[i].val = current;
