From 9756774948ed5364b52e3e797b92c40bbdd8aa07 Mon Sep 17 00:00:00 2001
From: Zebediah Figura <z.figura12@gmail.com>
Date: Mon, 17 Feb 2020 14:33:20 -0600
Subject: [PATCH] kernel32/tests: Add a test for mutex abandonment.

---
 dlls/kernel32/tests/sync.c | 21 ++++++++++++++++++++-
 1 file changed, 20 insertions(+), 1 deletion(-)

diff --git a/dlls/kernel32/tests/sync.c b/dlls/kernel32/tests/sync.c
index b4f5d1f1f0b..492cb1f72f2 100644
--- a/dlls/kernel32/tests/sync.c
+++ b/dlls/kernel32/tests/sync.c
@@ -193,6 +193,13 @@ static DWORD WINAPI mutex_thread( void *param )
     return 0;
 }
 
+static DWORD WINAPI abandon_mutex_thread( void *param )
+{
+    DWORD ret = WaitForSingleObject( mutex, 0 );
+    ok(!ret, "got %u\n", ret);
+    return 0;
+}
+
 static void test_mutex(void)
 {
     HANDLE thread;
@@ -379,12 +386,24 @@ todo_wine_if(getenv("WINEESYNC"))   /* XFAIL: due to the above */
     ret = ReleaseMutex( mutex2 );
     ok(ret, "got error %u\n", GetLastError());
 
+    thread = CreateThread( NULL, 0, abandon_mutex_thread, NULL, 0, NULL );
+    ret = WaitForSingleObject( thread, 2000 );
+    ok(ret == 0, "wait failed: %u\n", ret);
+
+    ret = WaitForSingleObject( mutex, 0 );
+    ok(ret == WAIT_ABANDONED, "got %u\n", ret);
+
+    ret = ReleaseMutex( mutex );
+    ok(ret, "got error %u\n", GetLastError());
+
+    ret = WaitForSingleObject( mutex, 0 );
+    ok(!ret, "got %u\n", ret);
+
     ret = CloseHandle( mutex );
     ok(ret, "got error %u\n", GetLastError());
 
     ret = CloseHandle( mutex2 );
     ok(ret, "got error %u\n", GetLastError());
-
 }
 
 static void test_slist(void)
